lib/model/receiveonlyfolder.go

```diff
@@ -0,0 +1,158 @@
+// Copyright (C) 2014 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package model
+
+import (
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/syncthing/syncthing/lib/config"
+	"github.com/syncthing/syncthing/lib/events"
+	"github.com/syncthing/syncthing/lib/fs"
+	"github.com/syncthing/syncthing/lib/protocol"
+	"github.com/syncthing/syncthing/lib/sync"
+	"github.com/syncthing/syncthing/lib/versioner"
+)
+
+type receiveOnlyFolder struct {
+	// WO folders are really just RW folders where we reject local changes...
+	sendReceiveFolder
+}
+
+func init() {
+	folderFactories[config.FolderTypeReceiveOnly] = newReceiveOnlyFolder
+}
+
+func newReceiveOnlyFolder(model *Model, cfg config.FolderConfiguration, ver versioner.Versioner, mtimeFS *fs.MtimeFS) service {
+	f := &receiveOnlyFolder{
+		sendReceiveFolder{
+			folder: folder{
+				stateTracker: newStateTracker(cfg.ID),
+				scan:         newFolderScanner(cfg),
+				stop:         make(chan struct{}),
+				model:        model,
+			},
+			FolderConfiguration: cfg,
+
+			mtimeFS:   mtimeFS,
+			dir:       cfg.Path(),
+			versioner: ver,
+
+			queue:       newJobQueue(),
+			pullTimer:   time.NewTimer(time.Second),
+			remoteIndex: make(chan struct{}, 1), // This needs to be 1-buffered so that we queue a notification if we're busy doing a pull when it comes.
+
+			errorsMut: sync.NewMutex(),
+
+			initialScanCompleted: make(chan struct{}),
+		},
+	}
+
+	f.configureCopiersAndPullers()
+
+	return f
+}
+
+func (f *receiveOnlyFolder) String() string {
+	return fmt.Sprintf("receiveOnlyFolder/%s@%p", f.folderID, f)
+}
+
+// validateAndUpdateLocalChanges reverts all local changes
+func (f *receiveOnlyFolder) validateAndUpdateLocalChanges(fs []protocol.FileInfo) []protocol.FileInfo {
+	fileDeletions := []protocol.FileInfo{}
+	dirDeletions := []protocol.FileInfo{}
+
+	for i, file := range fs {
+		if strings.Contains(file.Name, ".sync-conflict-") {
+			// This is a conflict copy, let's move on to the next file
+			continue
+		}
+
+		objType := "file"
+		action := "modified"
+		correctiveAction := "resync"
+
+		if file.IsDirectory() {
+			objType = "dir"
+		}
+
+		if file.IsDeleted() {
+			action = "deleted"
+		}
+
+		if len(file.Version.Counters) == 1 && file.Version.Counters[0].Value == 1 {
+			// A file, directory or symlink was added, which we'll have to remove again
+			action = "added"
+			if f.DeleteLocalChanges {
+				correctiveAction = "deleted"
+				if file.IsDirectory() {
+					dirDeletions = append(dirDeletions, file)
+				} else {
+					fileDeletions = append(fileDeletions, file)
+				}
+			} else {
+				correctiveAction = "none"
+			}
+		}
+		// Let's update the record to reflec that this is invalid and should be pulled again if possible
+		fs[i].Deleted = false
```

Comment:
I'm not convinced of the need for an "override all my changes, except for new files" mode, but I don't care strongly about that. By all means introduce whatever flags necessary to create such a mode if it's desired. But I do care very much about the default mode not undoing my changes, regardless if they are to existing files or new ones.